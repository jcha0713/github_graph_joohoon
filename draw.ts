/* prettier-ignore */
const drawings = {
  'J': [
    '   XXX',
    '     X',
    '     X',
    ' X   X',
    '  XXX '
  ],
  'O': [
    '  XXX ',
    ' X   X',
    ' X   X',
    ' X   X',
    '  XXX '
  ],
  'H': [
    ' X   X',
    ' X   X',
    ' XXXXX',
    ' X   X',
    ' X   X'
  ],
  'N': [
    ' X   X',
    ' XX  X',
    ' X X X',
    ' X  XX',
    ' X   X',
  ],
}

type alphabet = keyof typeof drawings

function drawWord(word: string) {
  const CANVAS_WIDTH = 51
  const WORD_WIDTH = 6
  const PADDING_X = Math.floor((CANVAS_WIDTH - WORD_WIDTH * word.length) / 2)
  const PADDING_Y = " ".repeat(CANVAS_WIDTH)

  const graph: string[][] = []

  graph.push(PADDING_Y.split(""))

  for (let row = 0; row < 5; row++) {
    let line = "".padStart(PADDING_X)

    for (const char of word.toUpperCase()) {
      const alphabet = char as alphabet
      if (drawings[alphabet]) {
        line += drawings[alphabet][row]
      }
    }
    graph.push(line.padEnd(CANVAS_WIDTH).split(""))
  }

  graph.push(PADDING_Y.split(""))

  const transposed = graph[0].map((_, col) => graph.map((row) => row[col]))
  return transposed.map((line) => line.join(""))
}

const pattern = drawWord("Joohoon")
const mask = pattern.flatMap((line) =>
  line.split("").map((char) => char === " ")
)

export { pattern, mask }

/*
                                                   
       XXX  XXX   XXX  X   X  XXX   XXX  X   X     
         X X   X X   X X   X X   X X   X XX  X     
         X X   X X   X XXXXX X   X X   X X X X     
     X   X X   X X   X X   X X   X X   X X  XX     
      XXX   XXX   XXX  X   X  XXX   XXX  X   X     
                                                   
*/
