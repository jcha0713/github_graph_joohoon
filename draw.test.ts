import { describe, expect, test } from "bun:test"
import { pattern, mask } from "./draw"

describe("drawWord", () => {
  test("WRITE MY NAME", () => {
    expect(pattern).toMatchSnapshot()
  })

  test("first X is at index 39", () => {
    for (let i = 0; i < 39; i++) {
      expect(mask[i]).toBe(true)
    }
    expect(mask[39]).toBe(false)
  })

  test("last X is at index 320", () => {
    for (let i = mask.length - 1; i > 320; i--) {
      expect(mask[i]).toBe(true)
    }
    expect(mask[320]).toBe(false)
  })

  test("mask has correct length of output", () => {
    const DAYS_IN_YEAR = 365
    const DAYS_ARE_NOT_IN_FULL_WEEK = 8
    expect(mask.length).toBe(DAYS_IN_YEAR - DAYS_ARE_NOT_IN_FULL_WEEK)
  })
})
