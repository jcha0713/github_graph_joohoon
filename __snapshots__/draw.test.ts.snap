// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`drawWord draws a word 1`] = `
[
  "       ",
  "       ",
  "       ",
  "       ",
  "       ",
  "    X  ",
  "     X ",
  " X   X ",
  " X   X ",
  " XXXX  ",
  "       ",
  "  XXX  ",
  " X   X ",
  " X   X ",
  " X   X ",
  "  XXX  ",
  "       ",
  "  XXX  ",
  " X   X ",
  " X   X ",
  " X   X ",
  "  XXX  ",
  "       ",
  " XXXXX ",
  "   X   ",
  "   X   ",
  "   X   ",
  " XXXXX ",
  "       ",
  "  XXX  ",
  " X   X ",
  " X   X ",
  " X   X ",
  "  XXX  ",
  "       ",
  "  XXX  ",
  " X   X ",
  " X   X ",
  " X   X ",
  "  XXX  ",
  "       ",
  " XXXXX ",
  "  X    ",
  "   X   ",
  "    X  ",
  " XXXXX ",
  "       ",
  "       ",
  "       ",
  "       ",
  "       ",
]
`;
